### DISCUSSION TITLE
[GsoC 2026 ] Agentic GraphRAG #36 on STaRK-MAG with OpenVINO — Working System + draft Proposal

### DISCUSSION THREAD

**Discussion by 1Suryansh1** [openvinotoolkit/openvino#34801] [Google Summer of Code]
Hi @bharagha @14pankaj,

I'm Suryansh Verma , B,Tech CSE Student from Delhi Technological University. I have a working, benchmarked GraphRAG system running on Intel OpenVINO, and I'd like to share it as my warm-up contribution for Project 36. 

My Laptop is intel AI PC with Intel Core Ultra 7 155H with integrated Intel Arc Graphics and NPU — all inference runs on-device via OpenVINO 

## Results

I evaluated the current system on the STaRK-MAG human-generated split with  FULL ALL 84 queries in a fully local, zero-shot setting on 1,872,968 NODES . The current no explicit rerank run achieves Hit@5 = 53.57 with average latency of about 10.2 seconds per query. 
 
human_eval is considered tougher than synthetic split  and synthetic has thousands of queries. 

| System | Hit@5 |
|--------|-------|
| BM25 | 41.67 |
| ada-002 | 41.67 |
| GritLM-7b | 44.04 |
| Claude3 Reranker | 45.24 |
| GPT4 Reranker | 46.43 |
| **This system** | **53.57** |

 the exact numbers here  https://arxiv.org/html/2404.13207v3
A few points matter here:

- I have considered Hit@5 bcoz this tool task is to retrieve top 5-10 which will be given to other agent. 
- My current system already outperforms the published GPT4 Reranker baseline on the same split , both on score and latency as avg latency is 23.43 sec as mentioned in paper and my system latency is 10.2 sec per query 
- I see this as a strong result for a system that is fully local, zero-shot, API-free, and not finetuned.
- Newer frontier systems such as HybGRAG and GraphRAFT report higher STaRK-MAG scores on synthetic split , but they come from later evaluation settings, and GraphRAFT is explicitly a finetuned method rather than the same local zero-shot regime. and their VRAM requirements are higher and some involves GNN also. 
- My goal in Project 36 is to close part of that gap with SLM-based anchor extraction, reflection, and better seed quality, while preserving the local OpenVINO deployment path. 
- I have done ablations and updating my repo to further strength my direction.

## Current System

I built a graph retrieval tool for agents, implemented and benchmarked on STaRK-MAG. The system is designed as a reusable retrieval backend for agentic workflows, not just a benchmark script.

**Offline setup**
- I precompute 384-dimensional OpenVINO embeddings for STaRK-MAG nodes using a bi-encoder.
- I store node text and embeddings in Neo4j with both vector and full-text indexes for hybrid seed retrieval.
- I also build a local CSR-style adjacency index for fast in-process neighbor lookup during traversal.

**Online retrieval flow**
- A query first goes through hybrid seed retrieval in Neo4j using dense retrieval, BM25/full-text, and Reciprocal Rank Fusion.
- The selected seeds are then expanded with an MCTS-based traversal engine that performs multi-hop exploration, backpropagation, and pruning.
- Candidate paths are scored with a trajectory-aware OpenVINO cross-encoder PRM that uses a sliding-window serialization of recent hops, which helps keep longer paths compatible with the model context window.
- Neighbor expansion uses degree-penalized semantic selection, which reduces hub-node domination during traversal.
- Results from multiple seed-specific search trees are merged into one ranked candidate list with trajectory evidence.

**Engineering choices**
- The runtime uses two graph layers: Neo4j for indexed seed retrieval and a in-memory CSR/NumPy graph store for fast local traversal.
- The entire query loop runs fully locally, with zero network calls during retrieval and graph search.

**OpenVINO implementation**
- The reranker supports NPU → GPU → CPU automatic fallback with recompilation, applies INT8 quantization on NPU , further indepth details could be found in repo

## Project 36 Direction

My proposal is to package this retrieval core as a **GraphQueryTool** inside a LangGraph agentic pipeline. The traversal engine remains a deterministic retrieval tool in the loop, while OpenVINO SLMs handle query understanding, reflection, and answer synthesis.

| Stage | Model | Device | Role |
|-------|-------|--------|------|
| Anchor SLM | Qwen3-1.7B INT4 | NPU | Schema-free entity extraction into structured JSON anchors |
| GraphQueryTool | Existing system | GPU | Anchor-primed seed retrieval, graph traversal, ranked candidates, evidence paths |
| Reflective SLM | Qwen3-4B INT4 | GPU | Retrieval sufficiency check and retry routing with better anchoring |
| Generative SLM | Qwen3-7B INT4 | GPU | Final answer synthesis from top candidates with citations if possible |

Text-to-Cypher fails silently on schemas — the LLM hallucinates relationship names and returns zero results with no recovery. My Anchor SLM extracts schema-free JSON entities instead. It never generates Cypher. Schema knowledge lives in the retrieval layer, not the LLM. 

## Project 37 compatibility  
( with @naitik-2006  fine tuning and fulfilling its requirement could be done , I have connected with him via linkedin , we can collaborate together towards converging solution with your guidance )
The three SLM layers can be fine tuned properly. 

The graph traversal engine is deterministic — no fine-tuning required, keeping the gradient signal clean and targeted. 

The same structure is also naturally extensible to multimodal retrieval because the seed-retrieval layer can ingest new embedding types without changing the traversal core.

## Conclusion

I'd like to propose this benchmarked implementation as my warm-up contribution for Project 36. It already demonstrates working Neo4j integration, OpenVINO inference, million-node-scale graph traversal, and retrieval evaluation on STaRK-MAG, which I believe maps directly to the goals of the project.

The choice of model for SLM could be further decided and extended to support multimodal nature also. 
I'd be happy to open a PR integrating this direction into the edge-ai-libraries ChatQnA pipeline.

Your suggestions and directions would be very helpful before I draft my final proposal

the image shows the PRESENT SYSTEM which would act as deterministic tool 
<img width="5851" height="9324" alt="mermaid-diagram-2026-03-19T15-40-50" src="https://github.com/user-attachments/assets/12937e16-62be-4687-aeb2-911a1242172d" />

the image below shows the proposed architecture 
<img width="5543" height="9168" alt="mermaid-diagram-2026-03-19T15-47-21" src="https://github.com/user-attachments/assets/f0a321af-6bcd-4aa1-8925-0e2c34120222" />

For Further clarification plz have a look at my architecture images 

**GitHub:** https://github.com/1Suryansh1/graphRAG 
I have added the mentors as collaborator and to guide for next steps and corrections ...

Suryansh Verma  
B.Tech CSE
Delhi Technological University

**Comment by 1Suryansh1** [comment]
Hi @bharagha @14pankaj,

Just a quick update on my post above! I realized the private repo invite as collaborator was probably a hassle to deal with, so I just made the repository fully public. You can check out here my working system : https://github.com/1Suryansh1/graphRAG

I'm drafting my final GSoC proposal around the architecture I shared. If you have a few minutes this week, any feedback on the pipeline would be really helpful to make sure I'm aligned with Project 36.

**Comment by bharagha** [comment]
Hi @1Suryansh1 

Apologies on the delayed response. Short answer: We are looking to engage on the specific problem statement we provided which is to extend the ChatQnA application with GraphRAG capability. We are not looking to bring in a separate GraphRAG based solution. As I see, it is also not aligned with the intention of this GSoC program. Hence my request is to revisit the problem statement and submit a proposal on how you will extend ChatQnA with GraphRAG capability that is able to handle noisy datasets which are also live as in new information is getting added on a continuous basis.

**Reply by 1Suryansh1** [reply]
Hi @bharagha 

Thank you for the clarification. I revisited the problem statement and updated my proposal to focus specifically on extending ChatQnA with GraphRAG capability, instead of proposing a separate GraphRAG solution. The revised design keeps the existing ChatQnA pipeline intact and adds an incremental, bounded GraphRAG extension intended for noisy and continuously updated datasets. I have now submitted the corrected proposal. 

Thank you again for the guidance.

**Comment by RahulSelvakumar** [comment]
Subject: GSoC 2026 Proposal Approach: Extending ChatQnA with Agentic GraphRAG  - Rahul Selvakumar

Message:

Hi @1Suryansh1  @4pankaj

I am finalizing my proposal for the 175-hour Agentic GraphRAG project and wanted to share my architectural approach to ensure it aligns with your vision for the existing application.

My primary focus is strictly on extending the current ChatQnA application rather than building a parallel solution.

Coming from a full-time background in automation testing, I place a heavy emphasis on data validation and pipeline stability. Additionally, recently deploying a live-data AI agent (apextracker via Google Cloud Run) gave me practical experience managing dynamic API ingestion. I plan to apply both of these experiences to address the complexities of real-world RAG systems.

**High-Level Extension Plan:**

**Continuous Data Ingestion:** Instead of treating the GraphDB as a static store, I am designing the pipeline to handle live, continuous data updates, ensuring the agent retrieves the most current state of the information.

**Noise Reduction & Fallback:** Before updating the Graph or vector embeddings, I will implement a pre-processing layer to clean and validate "noisy" datasets (like VIINA). If the GraphRAG retrieval confidence is low due to noise, the agent will gracefully fall back to the existing semantic similarity search.

**Agentic Routing within ChatQnA:** Integrate an agentic framework  directly into the ChatQnA flow, managing the logic while strictly keeping the selected local LLM under the 13B parameter limit.

**UI & Containerization:** Update the existing UI to expose the agent's multi-hop reasoning, and deploy the extended solution using Docker and Helm charts.

**A Quick Question:**
As I map out the continuous ingestion pipeline, do you have a preferred Graph Database that you feel integrates most smoothly with the existing ChatQnA architecture on Intel Core platforms?

Thank you for your time and guidance!

Best regards,
Rahul Selvakumar
[[GitHub Profile]](https://github.com/RahulSelvakumar)
Recent AI Agent project: [apextracker-1040501010782.us-central1.run.app/](https://apextracker-1040501010782.us-central1.run.app/)
